From d86366db6d485aa1b68e410954a238061ddb7eda Mon Sep 17 00:00:00 2001
Message-ID: <d86366db6d485aa1b68e410954a238061ddb7eda.1685027580.git.git@grubix.eu>
From: Michael J Gruber <git@grubix.eu>
Date: Tue, 15 Feb 2022 17:38:02 +0100
Subject: [PATCH 1/2] test: allow to use --full-scan

Some build infrastructure appears to have problems with mtime/stat,
leading to spurious failures (noticed on s390x and aarch64 with epel8).
Allow the test suite to be run with --full-scan so that release builds
can use the test suite while avoiding spurious failures.
---
 test/README             | 8 ++++++++
 test/test-lib-common.sh | 7 ++++++-
 test/test-lib.sh        | 2 +-
 3 files changed, 15 insertions(+), 2 deletions(-)

diff --git a/test/README b/test/README
index 10f127cb..ea540335 100644
--- a/test/README
+++ b/test/README
@@ -110,6 +110,14 @@ printed on screen. This printing can be disabled by setting the
 NOTMUCH_TEST_QUIET variable to a non-null value. Message on test
 failures and skips are still printed.
 
+File system/mtime issues
+------------------------
+Normally, `notmuch new` uses directory modification times, and the test
+suite checks that this works. If you encounter spurious test failures
+where messages added by one subtest appear to show up in the next subtest
+only (indicating mtime/stat problems) you can set NOTMUCH_TEST_FULLSCAN
+so that the test suite uses the --full-scan option of notmuch new.
+
 Skipping Tests
 --------------
 If, for any reason, you need to skip one or more tests, you can do so
diff --git a/test/test-lib-common.sh b/test/test-lib-common.sh
index 18fa29c0..0d64b771 100644
--- a/test/test-lib-common.sh
+++ b/test/test-lib-common.sh
@@ -29,6 +29,11 @@ if [[ -z "$NOTMUCH_SRCDIR" ]] || [[ -z "$NOTMUCH_BUILDDIR" ]]; then
 	exit 1
 fi
 
+if test -n "$NOTMUCH_TEST_FULLSCAN"
+then
+	NOTMUCH_NEW_OPTIONS="--full-scan"
+fi
+
 # Explicitly require external prerequisite.  Useful when binary is
 # called indirectly (e.g. from emacs).
 # Returns success if dependency is available, failure otherwise.
@@ -240,7 +245,7 @@ EOF
 # are also supported here, so see that function for details.
 add_message () {
     generate_message "$@" &&
-    notmuch new > /dev/null
+    notmuch new $NOTMUCH_NEW_OPTIONS > /dev/null
 }
 
 if test -n "$valgrind"
diff --git a/test/test-lib.sh b/test/test-lib.sh
index 1a6525df..569142e1 100644
--- a/test/test-lib.sh
+++ b/test/test-lib.sh
@@ -468,7 +468,7 @@ test_json_nodes () {
 }
 
 NOTMUCH_NEW () {
-    notmuch new "${@}" | grep -v -E -e '^Processed [0-9]*( total)? file|Found [0-9]* total file'
+    notmuch new $NOTMUCH_NEW_OPTIONS "${@}" | grep -v -E -e '^Processed [0-9]*( total)? file|Found [0-9]* total file'
 }
 
 NOTMUCH_DUMP_TAGS () {
-- 
2.41.0.rc1.178.gd8fba8690d

